### PR TITLE
Enable training strategy for Indexer

### PR DESCRIPTION
**Pull request by RissyRan**
# Description

Enable selective parameter training strategy for DeeSeek V3.2 Indexer - [paper](https://screenshot.googleplex.com/BHqSonnvgEw9R4a)
* Dense warm up stage:
   * Add `trainable_parameters_mask` flag, allowing specific parameters to be targeted for training while freezing the rest of the model.
   * Add `TrainableParametersMaskTest` unit tests for validation.
* Sparse training stage:
  * Update `indexer_sparse_training` flag to indicate Dense Warm-up stage or Sparse Training stage for DS v3.2. 
  * Add `test_indexer_gradients` unit test to verify proper gradient isolation.
* Renaming flags to avoid confusion
  * `use_sparse_indexer` --> `use_indexer`; `index_head_dim` --> `indexer_head_dim`; `index_n_heads` --> `indexer_n_heads`, and `index_topk` --> `indexer_topk`

# Tests

* Expect added unit tests are all green
* End-to-end functional - [logs](https://paste.googleplex.com/5503321424134144)
* Sanity check deepseek32_vs_reference_test - [link](https://paste.googleplex.com/4984126433263616), same as b/491486716

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/layers/attention_mla.py [hunk 1]
```diff
@@ -256,22 +256,50 @@ def __call__(
...
+      # Backward Pass (Main Model):
+      # - The main model optimizes its weights based solely on the LM loss.
+      # - The `indexer_mask` in the Attention layer prevents gradients from the main
+      #   loss from flowing into the Indexer's weights.
+      #
+      # Backward Pass (Indexer):
+      # - Gradients from the `indexer_loss` flow back to update the Indexer's weights.
+      # - The `stop_gradient` applied to the inputs acts as a mathematical wall, dropping
+      #   gradients to 0.0 and preventing the Indexer loss from altering the main model's
+      #   earlier layers.
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply src/maxtext/layers/attention_mla.py]: Make sense. I have added following code block and update the test accordingly

```
    if sparse_loss:
      # Detach main model components from the computational graph.
      # The indexer should match the main model, but the main model should not be influenced
      # by the indexer's learning progress via this loss in sparse training stage.
      query = jax.lax.stop_gradient(query)
      key = jax.lax.stop_gradient(key)
```

File: src/maxtext/optimizers/optimizers.py [hunk 2]
```diff
@@ -92,10 +96,19 @@ def get_optimizer(config, learning_rate_schedule, model=None):
...
-    return muon(**muon_kwargs)
+    base_opt = muon(**muon_kwargs)
   else:
     raise ValueError(f"{config.opt_type=} is not a supported.")
 
+  # If a whitelist of trainable parameters is provided, freeze everything else.
+  # When trainable_parameters_mask is empty, freeze_mask_fn is None and all parameters are trained.
+  trainable_patterns = getattr(config, "trainable_parameters_mask", None)
+  freeze_mask_fn = _get_path_mask_fn(trainable_patterns, match_returns_true=False)
+  if freeze_mask_fn is not None:
```
[the remark this answers is not in the document]
Comment (RissyRan) [reply src/maxtext/optimizers/optimizers.py:107]: Valid point. I update it accordingly.

File: src/maxtext/layers/attention_mla.py [hunk 3]
```diff
@@ -256,22 +256,50 @@ def __call__(
...
+      inputs_q_for_indexer = jax.lax.stop_gradient(inputs_q)
```
Comment (shuningjin) [inline src/maxtext/layers/attention_mla.py]: nit: can keep the names?

File: tests/unit/deepseek32_vs_reference_test.py [hunk 4]
```diff
@@ -458,14 +458,14 @@ def rotate_activation(x: torch.Tensor) -> torch.Tensor:
 
 class Indexer(torch.nn.Module):  # pylint: disable=missing-class-docstring
 
-  def __init__(self, args: ModelArgs, index_topk: int = 4):
+  def __init__(self, args: ModelArgs, indexer_topk: int = 4):
```
Comment (shuningjin) [inline tests/unit/deepseek32_vs_reference_test.py]: nit: keep the reference code unchanged in this file?

File: tests/unit/flop_calculation_test.py [hunk 5]
```diff
@@ -467,11 +467,10 @@ def test_deepseek32_671b_flops(self):
         "v_head_dim": 128,
         "skip_jax_distributed_system": True,
         # Indexer for DeepSeek Sparse Attention
-        "use_sparse_indexer": True,
-        "index_n_heads": 64,
-        "index_head_dim": 128,
-        "index_topk": 2048,
-        # TODO(ranran): remove after flash attention is supported
```
Comment (shuningjin) [inline tests/unit/flop_calculation_test.py:474]: can update attention according to this todo?

File: tests/unit/train_compile_test.py [hunk 6]
```diff
@@ -800,6 +800,54 @@ def test_deepseek32(self):
...
+            "model_name=deepseek-custom",
+            "per_device_batch_size=4",
+            "scan_layers=True",
+            "max_target_length=1024",
+            "attention=flash",
+            "use_tokamax_splash=True",
+            "engram_layers=[]",
+            # dense warmup
+            "indexer_sparse_training=False",
+            "indexer_loss_scaling_factor=0.1",
```
Comment (shuningjin) [inline tests/unit/train_compile_test.py:823]: add `trainable_parameters_mask=['.*indexer.*']`?

File: src/maxtext/configs/base.yml [hunk 7]
```diff
@@ -356,15 +356,15 @@ moba_topk: 8
...
-index_topk: 2048
-# Determines the token selection strategy for indexer loss: 
-# - False: Uses all tokens (Dense Warm-up).
-# - True: Uses only top-k tokens (Sparse Training).
+use_indexer: False
+indexer_head_dim: 128
+indexer_n_heads: 64
+indexer_topk: 2048
+# Determines the training strategy for the indexer:
+# - False (Dense Warm-up): Computes indexer loss over all tokens. Used with `trainable_parameters_mask` to freeze other model parameters.
```
Comment (shuningjin) [inline src/maxtext/configs/base.yml:364]: A suggestion on dense-warmup indexer-only training.

(a) Current approach: seems viable, calculate indexer loss + LM loss, update indexer only via optimizer masking
```
# forward
a1 -> indexer1 (indexer loss1) -> topk -> a2 -> indexer2  (indexer loss2)  -> topk -> a3 -> embedding -> LM loss
# backward
grad = (indexer loss1 + indexer loss2 + LM loss).backward()
# optimizer update, only track optimizer state for indexer1 and indexer2 and update param
masked_optimizer.update(params)
```
(b) I imagine the following might improve efficiency. Calculate indexer loss only, detach indexer input as in sparse training stage, normal optimizer would give correct update (yet optimizer masking could still eliminate redundant optimizer state)
```
# forward, eliminate final embedding and LM loss
a1.detach() -> indexer1 (indexer loss1) -> topk -> a2.detach() -> indexer2  (indexer loss2)  -> topk -> a3.detach() 
# backward, eliminate some gradient backprop due to detached graph
grad = (indexer loss1 + indexer loss2).backward()
# optimizer update, only track optimizer state for indexer1 and indexer2 and update param (same as before, masking is optional but recommend)
masked_optimizer.update(params)
```
For implementation, (1) the stop gradient would be same as sparse phase, can apply unconditionally. this alone might help backward during `jax.value_and_grad`. (2) not sure if it is easy to bypass final embedding and LM loss in forward pass, can be optional

File: src/maxtext/layers/attention_mla.py [hunk 8]
```diff
@@ -989,6 +1017,13 @@ def calculate_indexer_loss(
     Returns:
       The computed KL divergence loss.
     """
+    if sparse_loss:
```
Comment (shuningjin) [inline src/maxtext/layers/attention_mla.py]: can stop gradient unconditionally? (stated in the above)

File: src/maxtext/layers/attention_mla.py [hunk 9]
```diff
@@ -256,22 +256,50 @@ def __call__(
...
     # NOTE: If sequence length <= topk, indexer always selects all tokens.
-    if self.config.max_target_length <= self.index_topk:
+    if self.config.max_target_length <= self.indexer_topk:
       return None, None, None
 
     bsz, seqlen, _ = inputs_q.shape  # s = t = seqlen
 
     # Query Processing: Project from Latent low_rank_q
-    q = self.wq_b(low_rank_q)  # [b, t, q_lora_rank] -> [b, t, h * d]
+    if self.config.indexer_sparse_training:
```
Comment (shuningjin) [inline src/maxtext/layers/attention_mla.py]: can stop gradient unconditionally? (stated in the above)

File: tests/unit/flop_calculation_test.py (hunk 5 above)
Comment (RissyRan) [reply tests/unit/flop_calculation_test.py:474]: Sure, let's stick with Flash Attention. I initially thought the routing wouldn't matter for the FLOPs tests.

File: tests/unit/deepseek32_vs_reference_test.py (hunk 4 above)
Comment (RissyRan) [reply tests/unit/deepseek32_vs_reference_test.py]: Sure, will change back `indexer_topk` to `index_topk`.

File: src/maxtext/layers/attention_mla.py (hunk 3 above)
Comment (RissyRan) [reply src/maxtext/layers/attention_mla.py]: Sure, I could keep the name.

File: tests/unit/train_compile_test.py (hunk 6 above)
Comment (RissyRan) [reply tests/unit/train_compile_test.py:823]: Good catch! Let me add that.

File: src/maxtext/layers/attention_mla.py (hunk 9 above)
Comment (RissyRan) [reply src/maxtext/layers/attention_mla.py]: Discussed offline. We think it's no harm to apply stop_gradient for Dense warmup as well. Update remove.

File: src/maxtext/configs/base.yml (hunk 7 above)
Comment (RissyRan) [reply src/maxtext/configs/base.yml:364]: Addressed both (1) and (2). Thanks!

File: tests/unit/deepseek32_vs_reference_test.py [hunk 10]
```diff
@@ -461,9 +461,9 @@ class Indexer(torch.nn.Module):  # pylint: disable=missing-class-docstring
   def __init__(self, args: ModelArgs, index_topk: int = 4):
     super().__init__()
     self.dim: int = args.dim
-    self.n_heads: int = args.index_n_heads
-    self.n_local_heads = args.index_n_heads // world_size
-    self.head_dim: int = args.index_head_dim
+    self.n_heads: int = args.indexer_n_heads
+    self.n_local_heads = args.indexer_n_heads // world_size
+    self.head_dim: int = args.indexer_head_dim
```
Comment (shuningjin) [inline tests/unit/deepseek32_vs_reference_test.py]: nit: keep torch code above line 692 unchanged if possible?

https://github.com/AI-Hypercomputer/maxtext/blob/d3defd72c032949f1e1662c36e17cd3bec3e63f6/tests/unit/deepseek32_vs_reference_test.py#L692

Comment (RissyRan) [reply tests/unit/deepseek32_vs_reference_test.py]: OK, I see even those `ModelArgs` came from PT [reference](https://github.com/deepseek-ai/DeepSeek-V3.2-Exp/blob/87e509a2e5a100d221c97df52c6e8be7835f0057/inference/model.py#L18-L90). I will update it back.

File: src/maxtext/optimizers/optimizers.py [hunk 11]
```diff
@@ -24,31 +24,35 @@
...
     return None
 
-  compiled_patterns = [re.compile(pattern) for pattern in config.adamw_mask]
+  compiled_patterns = [re.compile(pattern) for pattern in patterns]
 
   def mask_fn(params):
-    def _is_decayed(path, _):
+    def _is_masked(path, _):
       # Join path keys into a single string for pattern matching (e.g., "layer1/bias")
       path_str = "/".join(str(getattr(p, "key", getattr(p, "idx", getattr(p, "name", p)))) for p in path)
```
Comment (Rohan-Bierneni) [inline src/maxtext/optimizers/optimizers.py]: nit: can replace with jax.tree_util.keystr(path)

Comment (RissyRan) [reply src/maxtext/optimizers/optimizers.py]: Thanks! Updated.
